Review

sadiuk opened the pull request.

sadiuk committed:
Added Pixel Abstraction

sadiuk committed:
Refactoring. Remade ImageData struct into Image class. Improved Loading/saving process.

sadiuk committed:
Added Gaussian pyramid reduce operation

sadiuk committed:
Added reduce example

sadiuk committed:
Refactoring. Added [[nodiscard]] to getters.

sadiuk committed:
Added Image expansion with gaussian pyramid. Added pyramid building

sadiuk committed:
Added [nodiscard]] for all functions, where the return value is important

sadiuk committed:
Replaced Pixel struct with alias of glm::u8vec4

sadiuk committed:
Some changes in image subtraction(unresolved issue with odd image width/height)

sadiuk committed:
Refactoring. Made GaussianPyramid an implementation of IPyramid interface

sadiuk committed:
Implemented Laplacian Pyramid

sadiuk committed:
Added some input images

mdavydov commented on src/retouch/Image/Pixel.h:
```diff
@@ -0,0 +1,32 @@
+                     static_cast<unsigned char>(m_green / divider),
+                     static_cast<unsigned char>(m_blue / divider),
+                     static_cast<unsigned char>(m_alpha)};
+    }
```
Якщо дані unsigned char, зазвичай, використовуєтсья clamp( int( op(x) ), 0, 255), а не приведення типу. Інакше -1 -> 255. В піраміді при різниці можуть бути від'ємні значення, тому ліпше використовувати short, a зайві біти використати для підвищення точності

mdavydov commented on src/retouch/Image/Pixel.h:
```diff
@@ -0,0 +1,32 @@
+    unsigned char m_red{0};
+    unsigned char m_green{0};
+    unsigned char m_blue{0};
+    unsigned char m_alpha{1};
```
непрозорий піксель = 255 для unsigned char

mdavydov commented on src/retouch/Image/Image.cpp:
```diff
@@ -0,0 +1,96 @@
+                 m_pixel_data(std::move(pixel_data)),
+                 m_width{width},
+                 m_height{height},
+                 m_bit_depth{bit_depth},
```
m_bit_depth більше схоже на m_pixel_size
//size_t buffer_size = m_width * m_height * m_bit_depth;

mdavydov commented on src/retouch/Image/Image.cpp:
```diff
@@ -0,0 +1,96 @@
+        unsigned char* image_buffer = nullptr;
+        try
+        {
+            image_buffer = new unsigned char[m_height * m_width * m_channels_count];
```
Тут вже m_bit_depth не враховується

mdavydov commented on src/retouch/Image/Image.cpp:
```diff
@@ -0,0 +1,96 @@
+        }
+        catch(std::bad_alloc&)
+        {
+            throw std::runtime_error("Unable to allocate memory for the image buffer.");
```
Тут обробляти цю помилку недоцільно, оскільки на цьому рівні невідомо, що з нею робити. Ліпше пропустити її на рівень виклику - там можна опрацювати bad_alloc і вивести якесь повідомлення, чи перемкнутися на інший підхід.

mdavydov commented on src/retouch/Image/Image.cpp:
```diff
@@ -0,0 +1,96 @@
+            m_bit_depth{bit_depth},
+            m_channels_count{channels_count}
+    {
+        size_t buffer_size = m_width * m_height * m_bit_depth;
```
m_bit_depth->m_channel_count*m_channel_size?
(або зробити шаблоном і написати sizeof(ChannelType) )

mdavydov commented on src/retouch/Image/Image.cpp:
```diff
@@ -0,0 +1,96 @@
+        }
+        catch(std::bad_alloc&)
+        {
+            throw std::runtime_error("Not enough heap memory for image allocation.");
```
Ліпше тут не обробляти

mdavydov commented on src/retouch/Image/Image.cpp:
```diff
@@ -0,0 +1,96 @@
+    }
+    Image Image::operator-(const Image &other) const
+    {
+        /*
```
Ліпше проблему різниці розміру покласти на того, хто викликає цей код, а тут зробити assert. Виклик різниці з зображеннями різного розміру - це помилка програмування, а не виняткова ситуація

mdavydov commented on src/retouch/Image/Image.cpp:
```diff
@@ -0,0 +1,96 @@
+            throw std::runtime_error("Properties of both images must be the same");
+        }
+        Image difference(m_width, m_height, m_bit_depth, m_channels_count);
+        for(int x = 0; x < m_width; x++)
```
Внутрішній цикл має бути по x, так буде швидше (менше промахів кеша)

mdavydov commented on src/retouch/Image/Image.h:
```diff
@@ -0,0 +1,68 @@
+              size_t bit_depth = KDefault_bit_depth,
+              size_t channels_count = KDefault_channels_count);
+
+        [[nodiscard]] const auto& getPixelData() const { return m_pixel_data; }
```
Як на мене, [[nodiscard]] тут зайвий, бо треба читати більше коду, а імовірність того, що хтось просто викличе getPixelData(), і не використає результат, мала.
[[nodiscard]] важливий там, те ми хочемо змусити того, хто викликає, обов'язково обробити рузультат. Напр
[[nodiscard]] bool invertMatrix(matrix& m); // вертає false, якщо матриця необоротня

Використання const auto& тут недоцільне, оскільки m_pixel_data - це std::unique_ptr, i неочевидно, який буде тип результату. Там, де можна вказати тип, ліпше не використовувати auto

mdavydov commented on src/retouch/Image/Image.h:
```diff
@@ -0,0 +1,68 @@
+
+        [[nodiscard]] Pixel getPixel(size_t x, size_t y) const
+        {
+                if(x >= m_width || y >= m_height) throw std::out_of_range("Index out of bounds of image data.");
```
Ліпше зробити assert, тоді в release буде швидше.
throw для помилок програмування доцільний при написанні інтерпретаторів

mdavydov commented on src/retouch/Pyramid/GaussianPyramid.cpp:
```diff
@@ -0,0 +1,89 @@
+
+    void GaussianPyramid::build()
+    {
+        while(m_layers.back().getHeight() / 2 > KMin_image_resolution &&
```
Частіше додають піксель, якщо розмір непарний, тобто  (h+1)/2

mdavydov commented on src/retouch/Pyramid/GaussianPyramid.cpp:
```diff
@@ -0,0 +1,89 @@
+                            original_image.getChannelsCount());
+
+        constexpr int KRadius = 2;
+        for(int x = 0; x < reduced_image.getWidth(); x++)
```
x - внутрішній цикл

mdavydov commented on src/retouch/Pyramid/LaplacianPyramid.cpp:
```diff
@@ -0,0 +1,18 @@
+
+    LaplacianPyramid::LaplacianPyramid(const Image &first_layer) : m_gaussian_pyramid(first_layer){}
+
+    void LaplacianPyramid::build()
```
Для піраміди важливо додати тест на незмінніть зображення після розкладу в піраміду і отриманні з піраміди початкового зображення

sadiuk committed:
Loads of refactoring.
1) Changed pixel operations. Now std::clamp is used instead of casting.
2) Pixels are now stored as glm::ivec16(i.e. short).
3) Set x to be in nested loop after y to hit cash more often when going through$
4) Removed m_bit_depth as a field of Image class.
5) Removed [[nodiscard]] in most cases.
6) Fixed issue with odd image size.
7) Other minor fixes and improvements.

sadiuk committed:
Added image addition

sadiuk committed:
Added image reconstruction using laplacian pyramid

mdavydov commented on src/retouch/Pyramid/LaplacianPyramid.cpp:
```diff
@@ -11,8 +11,55 @@ namespace retouch
+        {
+            for (int x = 0; x < expanded_image.getWidth(); x++)
+            {
+                std::vector<Pixel> neighbors;
```
Vector може довго працювати тут, простіше sum + weight

mdavydov commented on src/retouch/Pyramid/LaplacianPyramid.cpp:
```diff
@@ -11,8 +11,55 @@ namespace retouch
+            for (int x = 0; x < expanded_image.getWidth(); x++)
+            {
+                std::vector<Pixel> neighbors;
+                for(int i = -KRadius; i <= KRadius; i++)
+                {
+                    for(int j = -KRadius; j <= KRadius; j++)
```
Межі і крок ліпше наперед обчислити, щоб не робити if у циклі

mdavydov reviewed (commented):
Все ок, додав пару коментарів по швидкодії

sadiuk committed:
Optimized image reduction and expansion

sadiuk committed:
Removed IPyramid Interface and added empty Laplacian pyramid generation

sadiuk committed:
Implemented getting and setting subimages of an image

sadiuk committed:
Bud fixes in subimaging

sadiuk committed:
Implemented remap function

sadiuk committed:
Bug fixes and minor changes

sadiuk committed:
Implemented local laplacian filtering(buggy).

sadiuk committed:
Implemented Local laplacian filtering(now correct)

sadiuk committed:
Fixed bug with extra alpha power

sadiuk committed:
Implemented timer for benchmarking

sadiuk committed:
Separated image processing into parallel threads

sadiuk committed:
Fixed pyramid bugs

sadiuk committed:
Fixed shadows shifting to right-bottom side

sadiuk committed:
Replaced whole-pyramid building with only building one level
